src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -247,76 +373,105 @@ public SNIHandle CreateConnectionHandle(object callbackObject, string fullServer
-            string[] serverNamePartsByBackSlash = fullServerName.Split(BackSlashSeparator);
+            string hostName = null;
+            int port = -1;
+            Exception exception = null;
 
-            if (serverNamePartsByComma.Length < 2 && serverNamePartsByBackSlash.Length < 2)
+            if (string.IsNullOrWhiteSpace(fullServerName)) // when fullServerName is empty
             {
-                return new SNITCPHandle(fullServerName, DefaultSqlServerPort, timerExpire, callbackObject, parallel);
+                hostName = DefaultHostName;
+                port = DefaultSqlServerPort;
             }
-            else if (serverNamePartsByComma.Length == 2 && serverNamePartsByBackSlash.Length < 2)
+            else
             {
-                string hostName = serverNamePartsByComma[0];
-                int portNumber = -1;
-                try
+                string[] serverNamePartsByComma = fullServerName.Split(CommaSeparator);
+                string[] serverNamePartsByBackSlash = fullServerName.Split(BackSlashSeparator);
+
+                // when no port or instance name provided
+                if (serverNamePartsByComma.Length < 2 && serverNamePartsByBackSlash.Length < 2)
                 {
-                    portNumber = ushort.Parse(serverNamePartsByComma[1]);
-                    return new SNITCPHandle(hostName, portNumber, timerExpire, callbackObject, parallel);
+                    hostName = fullServerName;
+                    port = DefaultSqlServerPort;
                 }
-                catch (Exception e)
+                // when port is provided, and no instance name
+                else if (serverNamePartsByComma.Length == 2 && serverNamePartsByBackSlash.Length < 2)
                 {
-                    SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, SNICommon.InvalidConnStringError, e);
-                    return null;
+                    hostName = serverNamePartsByComma[0];
+                    string portString = serverNamePartsByComma[1];
+                    try
+                    {
+                        port = ushort.Parse(portString);
+                    }
+                    catch (Exception e)
+                    {
+                        exception = e;
+                    }
+                }
+                // when instance name is provided, and no port
+                else if (serverNamePartsByComma.Length < 2 && serverNamePartsByBackSlash.Length == 2)
+                {
+                    hostName = serverNamePartsByBackSlash[0];
+                    string instanceName = serverNamePartsByBackSlash[1];
+                    try
+                    {
+                        port = GetPortByInstanceName(hostName, instanceName);
+                    }
+                    catch (Exception e)
+                    {
+                        exception = e;
+                    }
                 }
```

Comment:
Yes, their length can be more than 2 which is wrong format of connection string fed by user.
In that case, hostname or port will have null or -1 respectively, and it will throw SNI error intentionally.